File: ClinicNow.Services/Database/Entities/MedicalDocument.cs
namespace ClinicNow.Services.Database.Entities;

/// <summary>
/// A file (PDF/image) plus an optional finding note attached to a patient's
/// record (CLAUDE.md §6: "MedicalDocument/MedicalFinding"). Legally-retained
/// health data - soft-delete only, never a hard DELETE. Uploaded by staff/a
/// doctor, downloaded by staff/doctors/administrators for any patient, or by
/// the patient themselves for their own record only (ownership enforced in
/// <c>MedicalDocumentService</c>, never trusted from the client).
///
/// The file bytes are validated against both the declared MIME type and the
/// file's actual magic bytes before being accepted (rulebook Part II §F:
/// "MIME + magic bytes, ne samo ekstenzija") - see
/// <see cref="MedicalDocumentService"/>.
/// </summary>
public class MedicalDocument : ISoftDelete
{
    public int Id { get; set; }

    public int PatientId { get; set; }
    public Patient Patient { get; set; } = null!;

    public string FileName { get; set; } = string.Empty;
    public string ContentType { get; set; } = string.Empty;
    public byte[] FileData { get; set; } = [];
    public long FileSizeBytes { get; set; }

    /// <summary>Free-text clinical finding/note attached alongside the file (may be empty if it's just a scan).</summary>
    public string? Description { get; set; }

    public int UploadedByUserId { get; set; }
    public User UploadedByUser { get; set; } = null!;

    public DateTime CreatedAtUtc { get; set; }

    public bool IsDeleted { get; set; }
    public DateTime? DeletedAtUtc { get; set; }
}

File: UI/clinicnow_desktop/lib/models/medical_document.dart
/// Mirrors the backend's `MedicalDocumentDto`. Never carries raw file bytes -
/// only [downloadUrl] (the actual bytes are served from a dedicated endpoint).
class MedicalDocument {
  final int id;
  final int patientId;
  final String patientName;
  final String fileName;
  final String contentType;
  final int fileSizeBytes;
  final String? description;
  final String uploadedByName;
  final DateTime createdAtUtc;
  final String downloadUrl;

  const MedicalDocument({
    required this.id,
    required this.patientId,
    required this.patientName,
    required this.fileName,
    required this.contentType,
    required this.fileSizeBytes,
    this.description,
    required this.uploadedByName,
    required this.createdAtUtc,
    required this.downloadUrl,
  });

  factory MedicalDocument.fromJson(Map<String, dynamic> json) {
    return MedicalDocument(
      id: json['id'] as int,
      patientId: json['patientId'] as int,
      patientName: json['patientName'] as String? ?? '',
      fileName: json['fileName'] as String? ?? '',
      contentType: json['contentType'] as String? ?? '',
      fileSizeBytes: json['fileSizeBytes'] as int? ?? 0,
      description: json['description'] as String?,
      uploadedByName: json['uploadedByName'] as String? ?? '',
      createdAtUtc: DateTime.parse(json['createdAtUtc'] as String),
      downloadUrl: json['downloadUrl'] as String? ?? '',
    );
  }
}
